### PR TITLE
Add support for repeated XML elements without a name attribute

### LINKED ISSUE #36541: Mixing JSON and XML configuration leads to subtle bugs when using the "Name" attribute
### Functional impact
In our test scenario, Serilog configuration failed to load properly because of this bug.

### Minimal repro steps

1. Create a .NET Core 2.0 console application with an appsettings.xml and an appsettings.json file
2. Fill in the appsettings.xml file like this:

```xml
<settings>
  <Serilog>
    <WriteTo Name="0">
      <Args>
        <pathFormat>C:\FooBar\Logs</pathFormat>
      </Args>
    </WriteTo>
  </Serilog>
</settings>
```

3. Fill in the appsettings.json like this:

```json
{
  "Serilog": {
    "WriteTo": [
      {
        "Name": "RollingFile"
      }
    ]
  }
}
```

4. Configure both files with "Build Action" set to "Content"
5. Update Program.cs to the following:

```csharp
  class Program
  {
    static void Main(string[] args)
    {
      var jsonConfiguration = new ConfigurationBuilder()
        .SetBasePath(Directory.GetCurrentDirectory())
        .AddJsonFile("appsettings.json", optional: false)
        .Build();
      var xmlConfiguration = new ConfigurationBuilder()
        .SetBasePath(Directory.GetCurrentDirectory())
        .AddXmlFile("appsettings.xml", optional: false)
        .Build();
      
      Console.WriteLine("*** JSON CONFIG ***");
      foreach (var (key, value) in jsonConfiguration.AsEnumerable())
      {
        Console.WriteLine($"{key, -25} = {value}");
      }
      Console.WriteLine();

      Console.WriteLine("*** XML CONFIG ***");
      foreach (var (key, value) in xmlConfiguration.AsEnumerable())
      {
        Console.WriteLine($"{key, -25} = {value}");
      }
      Console.WriteLine();

      Console.ReadLine();
    }
  }
```

6. Add the following NuGet references to your project

```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="2.0.1" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.0.1" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Xml" Version="2.0.1" />
  </ItemGroup>
```

7. Startup the application

### Expected result
*The XML config should NOT produce a 'Name' key in the configuration*  

I would expect the following output:

```
*** JSON CONFIG ***
Serilog                   =
Serilog:WriteTo           =
Serilog:WriteTo:0         =
Serilog:WriteTo:0:Name    = RollingFile

*** XML CONFIG ***
Serilog                   =
Serilog:WriteTo           =
Serilog:WriteTo:0         =
Serilog:WriteTo:0:Args    =
Serilog:WriteTo:0:Args:pathFormat = C:\FooBar\Logs
```

### Actual result
*The XML config produces a 'Name' key in the configuration*  

```
*** JSON CONFIG ***
Serilog                   =
Serilog:WriteTo           =
Serilog:WriteTo:0         =
Serilog:WriteTo:0:Name    = RollingFile

*** XML CONFIG ***
Serilog                   =
Serilog:WriteTo           =
Serilog:WriteTo:0         =
Serilog:WriteTo:0:Name    = 0
Serilog:WriteTo:0:Args    =
Serilog:WriteTo:0:Args:pathFormat = C:\FooBar\Logs
```

### Impact

As you can see, both the JSON and the XML file produce a key 'Serilog:WriteTo:0:Name', once with 0 and once with RollingFile. Due to the way the ConfigurationBuilder works, the _last_ key will win. 
When the value of Serilog:WriteTo:0:Name is 'RollingFile', Serilog will successfully configure the sink and configuration will succeed. When the value is '0' however, Serilog will skip configuring this sink.

This means that when the configuration builder is called like follows:

```csharp
      var configuration = new ConfigurationBuilder()
        .SetBasePath(Directory.GetCurrentDirectory())
        .AddJsonFile("appsettings.json", optional: false)
        .AddXmlFile("appsettings.xml", optional: false)
        .Build();
```

It will **not** work, because the key Serilog:WriteTo:0:Name from the XML file wins, breaking Serilog.

If you change it to this however:

```csharp
      var configuration = new ConfigurationBuilder()
        .SetBasePath(Directory.GetCurrentDirectory())
        .AddXmlFile("appsettings.xml", optional: false)
        .AddJsonFile("appsettings.json", optional: false)
        .Build();
```

The program will correctly configure Serilog.

### Further information

This issue is actually closely related to another issue I've created here: dotnet/extensions#745 
The proposed solution there also solves this issue, as it removes the need to use "Name" attributes to simulate arrays in XML and thus avoids this subtle problem.

### LINKED ISSUE #36561: Support repeated elements in XML configuration
### Functional impact
*Configuration that is possible from JSON is not always possible from XML, for example configuring Serilog sinks are configured as an array in JSON. Arrays in XML are not supported by the XmlConfigurationProvider.*

### Minimal repro steps

1. Create a .NET Core 2.0 application with an appsettings.xml file
2. Fill in the appsettings.xml file like this:

```xml
          <settings>
            <DefaultConnection>
                <ConnectionString>TestConnectionString1</ConnectionString>
                <Provider>SqlClient1</Provider>
            </DefaultConnection>
            <DefaultConnection>
                <ConnectionString>TestConnectionString2</ConnectionString>
                <Provider>SqlClient2</Provider>
            </DefaultConnection>
          </settings>
```

3. Load this file with the following snippet in Program.cs or Startup.cs

```csharp
new ConfigurationBuilder()
      .SetBasePath(Directory.GetCurrentDirectory())
      .AddXmlFile($"appsettings.xml")
      .Build();
```

4. Startup the application

### Expected result
*Unique keys should be produced for each entry in the XML settings*  

I would expect the following output:

```
"DefaultConnection:0:ConnectionString"="TestConnectionString1"
"DefaultConnection:0:Provider"="SqlClient1"
"DefaultConnection:1:ConnectionString"="TestConnectionString2"
"DefaultConnection:1:Provider"="SqlClient2"
```

### Actual result

```
System.FormatException occurred
  HResult=0x80131537
  Message=A duplicate key 'DefaultConnection:ConnectionString' was found.
  Source=<Cannot evaluate the exception source>
```

### Further technical details

I took the liberty of actually implementing something like this, the full source code is available at https://github.com/amoerie/configuration.xml

I've added extra unit tests to cover the various scenarios (turns out there's a magical 'Name' attribute with some special rules there) and all tests seem to be green. 

I refrained from immediately creating a pull request, because nobody (besides myself) was asking for this and it is actually a considerable refactoring of the existing code, so it seemed better to start a conversation first. Thank you for your consideration.

### PR DESCRIPTION
**Pull request by amoerie**
Fixes #36541
Fixes #36561 

This pull request adds support in Microsoft.Extensions.Configuration.Xml for repeated XML elements without requiring a Name attribute.
This solves a particularly subtle bug when configuring Serilog from an XML configuration source. For a full description, see https://github.com/dotnet/runtime/issues/36541

The original implementation of the XmlStreamConfigurationProvider has been modified to do the following:

- Maintain a stack of encountered XML elements while traversing the XML source. This is needed to detect siblings.
- When siblings are detected, automatically append an index to the generated configuration keys. This makes it work exactly the same as the JSON configuration provider with JSON arrays.

Tests are updated to reflect the new behavior:
- Tests that verified an exception occurs when entering duplicate keys have been removed. Duplicate keys are supported now.
- Add tests that verify duplicate keys result in the correct configuration keys, with all the lower/upper case variants and with support for the special "Name" attribute handling.

Thank you for your consideration.

### CODE REVIEW

**Comment by amoerie** [comment]
I could use a little help with the failing build, I don't see how it's related to my changes.

**Comment by maryamariyan** [comment]
The failing build 

```
  Discovering: Microsoft.Extensions.Hosting.Unit.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  Microsoft.Extensions.Hosting.Unit.Tests (found 84 test cases)
  Starting:    Microsoft.Extensions.Hosting.Unit.Tests (parallel test collections = on, max threads = 2)
    Microsoft.Extensions.Hosting.Internal.HostTests.BackgroundServiceAsyncExceptionGetsLogged [FAIL]
      'BackgroundService failed' did not get logged
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs(1256,0): at Microsoft.Extensions.Hosting.Internal.HostTests.BackgroundServiceAsyncExceptionGetsLogged()
[40m[32minfo[39m[22m[49m: Microsoft.Extensions.Hosting.Tests.HostTests[0]
      Request starting
  Finished:    Microsoft.Extensions.Hosting.Unit.Tests
=== TEST EXECUTION SUMMARY ===
   Microsoft.Extensions.Hosting.Unit.Tests  Total: 84, Errors: 0, Failed: 1, Skipped: 0, Time: 15.317s
```

is tracked by https://github.com/dotnet/runtime/issues/43389

**Comment by maryamariyan** [comment]
Thanks @amoerie for porting over this change to dotnet/runtime. 

We'll need to do a full review on the changes as it seems like there is a considerable change being introduced.
This changes has been upvoted by community and @ajcvickers mentioned in earlier PRs that the repo consolidations were part of the blocker for this fix to getting merged in.

Also adding @HaoK to this PR. Do you have any major hesitations on this enhancement getting introduced into extensions configuration or it sounds good to you?

**Comment by amoerie** [comment]
> We'll need to do a full review on the changes as it seems like there is a considerable change being introduced.

I just want to add a little note to this sentence: the internal changes are indeed considerable, but the external compatibility has not been broken. I did not modify any existing tests (except for 2 tests that verified duplicate keys are not supported - now they are) and only added tests to verify the new supported behavior. 

In any case, I'm very interested to hear your feedback on my code.

File: src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs [hunk 1]
```diff
@@ -42,57 +42,61 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
...
                 XmlNodeType preNodeType = reader.NodeType;
                 while (reader.Read())
                 {
                     switch (reader.NodeType)
                     {
                         case XmlNodeType.Element:
-                            prefixStack.Push(reader.LocalName);
-                            ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                            ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                            var parent = currentPath.Any() ? currentPath.Peek() : null;
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:57]: nit this line and some other instances: `We only use var when it's obvious what the variable type is (e.g. var stream = new FileStream(...) not var stream = OpenStandardInput()).`

taken from [coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)

File: src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs [hunk 2]
```diff
@@ -20,6 +20,16 @@ public override void Combine_after_other_provider()
             // Disabled test due to XML handling of empty section.
         }
 
+        public override void Load_from_single_provider_with_duplicates_throws()
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:23]: Should we be doing the same thing as below instead?
https://github.com/dotnet/runtime/blob/8fcb543625324c81f934e67a15077ad903dd3c9f/src/libraries/Microsoft.Extensions.Configuration.Json/tests/ConfigurationProviderJsonTest.cs#L13-L19

and 

https://github.com/dotnet/runtime/blob/8fcb543625324c81f934e67a15077ad903dd3c9f/src/libraries/Microsoft.Extensions.Configuration.CommandLine/tests/ConfigurationProviderCommandLineTest.cs#L38-L49

File: src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs (hunk 1 above)
Comment (amoerie) [reply src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs]: Ah I actually perused through the code style but I must have missed this. 
Updated!

File: src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs (hunk 2 above)
Comment (amoerie) [reply src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:23]: Well the behavior with duplicate keys is already covered by other tests. These tests in the JSON provider simply test if the provider does not crash on duplicate keys, while my added tests also verify if the actual output is correct.

But when in doubt, it's always better to be consistent. I've followed the JSON example and implemented these tests the same for the XML provider.

**Comment by HaoK** [comment]
I don't have any particular concerns, other than the general one around accidental breaking changes.  This behavior in particular has been asked for frequently, and it was always near the top of the list of PRs we were considering, so I think its a good candidate to be worth the risk.

**Comment by HaoK** [comment]
One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

**Comment by amoerie** [comment]
> One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

Are you saying there's other config provider implementations for XML files somewhere? Could you point me to them? I was operating under the assumption that everything would use the classes that I modified.

**Comment by amoerie** [comment]
Small update: I went digging a little and found no other implementations for XML files. In fact, I am convinced now that XML files also simply pass through the XmlStreamConfigurationProvider. See this class:

https://github.com/dotnet/runtime/blob/f3d84ca003943b09a8b80091e4b61221be339b22/src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationProvider.cs#L8-L29

**Comment by amoerie** [comment]
> /azp run runtime

Thanks @maryamariyan 

Perhaps the [PR guide](https://github.com/dotnet/runtime/blob/master/docs/pr-guide.md) could use an update, because I didn't see any option aside from pushing more commits or closing and reopening this PR.

**Comment by amoerie** [comment]
I'll close and reopen this PR to avoid merge commits.

**Comment by maryamariyan** [comment]
@amoerie how about the "rerun all checks" tab?
I think it refers to the "rerun" links in https://github.com/dotnet/runtime/pull/43722/checks?check_run_id=1302103237

**Comment by safern** [comment]
I believe that is because that is not available for contributors, only for collaborators. I believe write permissions are needed in order to use /azp run... and to see the rerun buttons on the checks tab.

**Comment by maryamariyan** [comment]
> One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

> Small update: I went digging a little and found no other implementations for XML files. In fact, I am convinced now that XML files also simply pass through the XmlStreamConfigurationProvider.

We have three file providers, and XmlConfigurationProvider is one of them, the other two are JsonConfigurationProvider and IniConfigurationProvider.

If I understood correctly @HaoK recommended if we are going to accept this breaking change in behavior that should also be done for the other two.

**Comment by HaoK** [comment]
Actually I wasn't talking about the other types of provider, I was referring to the old XmlConfigurationProvider, but indeed its just routing calls through the same code path so we should be good.

https://github.com/amoerie/runtime/blob/88e4f545e95d6952fbb34926f1dc9156b4e0baf6/src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationProvider.cs#L27

**Comment by amoerie** [comment]
I've recreated my pull request here: #44608

This cleans up the commit history (i've squashed my commits into one) and properly retries the CI.

**Pull request closed without merging**